src/coherence/ai.py

```diff
@@ -337,6 +337,87 @@ def __init__(self, extractor: Union[ValueExtractor[T, E], str], over_sampling_fa
         self.oversamplingFactor = over_sampling_factor
 
 
+@proxy("coherence.hnsw.HnswIndex")
+class HnswIndex(AbstractEvolvable):
+    DEFAULT_SPACE_NAME: Final[str] = "COSINE"
+    """The default index space name."""
+    DEFAULT_MAX_ELEMENTS: Final[int] = 4096
+    """
+    The default maximum number of elements the index can contain is 4096
+    but the index will grow automatically by doubling its capacity until it
+    reaches approximately 8m elements, at which point it will grow by 50%
+    whenever it gets full.
+    """
+    DEFAULT_M: Final[int] = 16
+    """
+    The default number of bidirectional links created for every new
+    element during construction is 2-100. Higher M work better on datasets
+    with high intrinsic dimensionality and/or high recall, while low M work
+    better for datasets with low intrinsic dimensionality and/or low recalls.
+    The parameter also determines the algorithm's memory consumption,
+    which is roughly M * 8-10 bytes per stored element. As an example for
+    dim=4 random vectors optimal M for search is somewhere around 6,
+    while for high dimensional datasets (word embeddings, good face
+    descriptors), higher M are required (e.g. M=48-64) for optimal
+    performance at high recall. The range M=12-48 is ok for the most of the
+    use cases. When M is changed one has to update the other parameters.
+    Nonetheless, ef and ef_construction parameters can be roughly estimated
+    by assuming that M*ef_{construction} is a constant. The default value is
+    16.
+    """
+    DEFAULT_EF_CONSTRUCTION: Final[int] = 200
+    """
+    The parameter has the same meaning as ef, which controls the
+    index_time/index_accuracy. Bigger ef_construction leads to longer
+    construction, but better index quality. At some point, increasing
+    ef_construction does not improve the quality of the index. One way to
+    check if the selection of ef_construction was ok is to measure a recall
+    for M nearest neighbor search when ef =ef_construction: if the recall is
+    lower than 0.9, than there is room for improvement. The default value is
+    200.
+    """
+    DEFAULT_EF_SEARCH: Final[int] = 50
+    """
+    The parameter controlling query time/accuracy trade-off. The default
+    value is 50.
+    """
+    DEFAULT_RANDOM_SEED: Final[int] = 100
+    """The default random seed used for the index."""
+
+    def __init__(
+        self,
+        extractor: Union[ValueExtractor[T, E], str],
+        dimensions: int,
+        space_name: str = DEFAULT_SPACE_NAME,
+        max_elements: int = DEFAULT_MAX_ELEMENTS,
+        m: int = DEFAULT_M,
+        ef_construction: int = DEFAULT_EF_CONSTRUCTION,
+        ef_search: int = DEFAULT_EF_SEARCH,
+        random_seed: int = DEFAULT_RANDOM_SEED,
+    ) -> None:
+        """
+
```

Comment:
No documentation here aside from the parameters.

Reply:
Added documentation